### ISSUE TITLE
GDScripts load threading undefined behavior.

### ISSUE THREAD

**Issue by darthLeviN** [godotengine/godot#98633]
### Tested versions

4.3, 3.5

### System information

Kubuntu 22.04

### Issue description

Just to note here to avoid confusion. The resources involved here are not manipulated after load. we're simply loading scripts for use.

**Issue 1**
This issue is about when you want to load script heavy projects. When threads are available, the loader is not using parallel threads to load independent scripts.

If you try to do it manually, if two scripts depend on another script, then you MIGHT get undefined behavior.
Loading the same resource from different threads, the issues will be different from platform to platform. Some might not reveal any problem. Some might freeze up.

One platform will be fine, while another might tell you that `GDScript` has no method called `new` !, while another one might just deadlock if you try to even load something from another thread.

I know that there is `load-threaded` that we can use in the ResourceLoader in godot 4 but that should serve mostly as a utility rather than avoiding the problem. Controlling what resource loads and where is very difficult in godot as it's done implicitly in a lot of places.

---------------
**Issue 2**
Another issue that exists is that when you are loading a script and it references 10 other scripts that are independent of each other, those 10 scripts don't seem to be threaded. even when you use `load_threaded_request` with `use_sub_threads` set to true, it seems to ignore the fact that other scripts are dependencies.

calling `get_dependencies` also returns an empty array for a `GDScript` while it references many scripts.

---------------
Suggestion :
- Scripts must be treated as regular resources. All referenced script paths must be cached with the editor to guide the load during runtime by knowing the dependency hierarchy. They should also be visible when calling `get_dependencies` on a `GDScript`.
- The `load` function should automatically thread for built in types that only have built in dependencies in the dependency hierarchy. Same should happen for any user defined type that it's script does not override `_init` so that loading it would have no side effects other than member variable initialization. We might want to extend this to user defined types that have some sort of script annotation set for themselves and all their dependencies. Example hirearchy :
    (GDScript -> preloaded Resource with no _init -> exported variable of a type which has a special annotation). 
- Possible deadlocks might happening. look out for deadlocks. I don't know how good the current threaded loading mechanism is but we need a proper scheduler.

### Steps to reproduce

- Try to load GDScripts that both depend on another script in parallel.
- Try to load a GDScript in another thread manually.
- Make sure you test on multiple platforms including html with threads enabled. For godot 3.5 html and native linux it told me there is no `new` in `GDScript`, for godot 4 in native linux everything is fine (so far, cannot do more testing at the time) but with html export in it freezes (after confriming threads do work with running a test function in a separate thread).
- Release html exports were tested as well

### Minimal reproduction project (MRP)

[scriptloadbug.zip](https://github.com/user-attachments/files/17554959/scriptloadbug.zip)
This is a godot 4 example. It freezes on html5 while confirming that thread support is working.
Let me know if you need more example projects. It's easier to reproduce in bigger projects.

**Comment by darthLeviN** [comment]
Seems like the issue of my test project was related to an infinite loop that javascript doesn't like.
I will close this issue for now till i face a problem in this area.
